### ISSUE TITLE
Request: Warn when ignored lines are actually reached

### ISSUE THREAD

**Issue by wchargin** [istanbuljs/nyc#919]
I use `// istanbul ignore next` and friends only to mark code that
should never be reached—for instance, internal assertion failures. In
the event that one of these ignored elements _is_ reached, this
indicates a bug in my code. Ideally, a unit test would catch this, but
there are cases when it might not (e.g., an expect-to-throw unit test
that throws for the wrong reason, or code that attempts to gracefully
recover from the unexpected condition).

It would be nice for the CLI or lcov-report or other interface to
indicate to the user when an ignored element is reached. Currently, it
is not even possible to discern this from the lcov-report: if a line is
marked `istanbul ignore if`, then it has no execution count annotation
even if it was executed.

The docs indicate that an accepted use of ignore-directives, at at least
one point in time, was to inflate test coverage numbers by marking
reachable code that people [simply don’t feel like testing][docs]. If
this is still the case, then this functionality should presumably raise
only a warning, or be opt-in/opt-out.

[docs]: https://github.com/gotwarlost/istanbul/blob/bc84c315271a5dd4d39bcefc5925cfb61a3d174a/ignoring-code-for-coverage.md#ignore-default-assignments

**Comment by coreyfarrell** [comment]
To be honest I'm not sure how this would be accomplished without major work to istanbul-lib-instrument.  In my own expect-to-throw tests I would test for a specific exception type so for example a test would only pass if a TypeError was thrown.  This is how I would differentiate between expected and unexpected exceptions.

I use the istanbul ignore directives to deal with situations that cannot be reasonably tested, usually where my code has to deal with an error that could be received from an external server.

**Comment by JaKXz** [comment]
If I may be so bold, it sounds like dead code is being kept around. My 2c would be to delete that code and if things break, you find your issue. 

I'm going to close this issue for now, since I don't foresee this being a maintainable addition to istanbul. Please correct me if I'm wrong.

**Comment by wchargin** [comment]
Hi @JaKXz—thanks for your response.

> If I may be so bold, it sounds like dead code is being kept around. My
> 2c would be to delete that code and if things break, you find your
> issue.

No, I definitely disagree with this statement. Assertions that verify
inline that the code is correct can be immensely valuable. For instance,
the implementation of CPython’s TimSort algorithm has 58 assert
statements by my count, which, assuming that the code is correct, can
not be hit no matter what inputs are provided. This does not make them
any less valuable: they help readers understand the code.

When Josh Bloch ported TimSort to the Java standard libraries,
these assertions were critical in helping him understand the code and
gain confidence that his implementation was correct. He says that even
with his unit tests, “Tim Peters' assertions were instrumental in
finding and fixing at least one bug in [his] port that would otherwise
have gone undetected”.

Here is a simple, self-contained example. Suppose that we implement a
binary search function, which includes an invariant check:

```javascript
/** Return an index of `target` in sorted `arr`, or `-1` if none exists. */
function binarySearch(arr, target) {
    let lo = 0;  // lower bound, inclusive
    let hi = arr.length;  // upper bound, exclusive
    while (lo < hi) {
        const mid = lo + Math.floor((hi - lo) / 2);
        const here = arr[mid];
        if (here === target) {
            return mid;
        } else if (here < target) {
            lo = mid + 1;
        } else if (here > target) {
            hi = mid - 1;
        }
        // istanbul ignore next
        if (lo > hi) {
            throw new Error("Invariant violation!");
        }
    }
    return -1;
}
```

If the implementation is correct, that `if`-statement will never be
reached. This is a mathematical property: it’s not like there’s some
dependency that we can mock out to simulate a failure at a lower level
of the stack. And, indeed, we might write some small test cases and see
that they pass. Following your advice, we might remove the “dead” check.

But that implementation is in fact incorrect, and the invariant can be
hit on some slightly larger inputs: `([2, 4, 6], 1)` returns the right
answer, but `([2, 4, 6, 8], 1)` throws!

Now, we might have some code that exercises this, but catches the error:

```javascript
try {
    loadPlugin(foo);  // transitively calls `binarySearch`
} catch (e) {
    useFallbackBehavior();
}
```

If the assertion were still around, Istanbul would be able to tell that
it was hit, and alert us that our invariant had been violated, even
though the calling code itself did not fail because the exception was
caught. Removing the assertion only makes the code less robust.

Does this make sense?

**Comment by JaKXz** [comment]
I think I see what you're saying @wchargin. It would be cool to replace `//istanbul ignore next` with `//istanbul warn next`.

`//istanbul warn next` could do the same thing as `//istanbul ignore next` (not count that line towards the coverage totals) but also throw a warning if coverage is collected.

**Comment by wchargin** [comment]
Yes: something like `//istanbul warn` or `//istanbul unreachable` would
be great!

**Comment by wchargin** [comment]
@stale[bot]: This issue has not been resolved.

**Comment by coreyfarrell** [comment]
I'd added a label to prevent stale bot from touching this issue in the future.

**Comment by wchargin** [comment]
Thank you!
